src/Facility.CodeGen.CSharp/CSharpGenerator.cs

```diff
@@ -356,11 +375,80 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
 									}
 								}
 
-								if (validatableFields.Count != 0)
+								if (validateFields.Count != 0)
+								{
+									code.WriteLineSkipOnce();
+									foreach (var fieldInfo in validateFields)
+									{
+										var type = context.GetFieldType(fieldInfo);
+										var propertyName = context.GetFieldPropertyName(fieldInfo);
+										var validation = fieldInfo.Validation!;
+
+										switch (type.Kind)
+										{
+											case ServiceTypeKind.Enum:
+											{
+												code.WriteLine($"if (!{propertyName}.IsDefined())");
+												using (code.Indent())
+													code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\");");
+
+												break;
+											}
+
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+													GenerateRangeCheck(code, $"{propertyName}.Length", validRange);
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													code.WriteLine($"if (!System.Text.RegularExpressions.Regex.IsMatch({propertyName}, \"{validPattern}\")");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\");");
+												}
+
+												break;
+											}
+
+											case ServiceTypeKind.Int32:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!);
+												break;
+											}
+											case ServiceTypeKind.Double:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!, "D");
+												break;
+											}
+											case ServiceTypeKind.Int64:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!, "L");
+												break;
+											}
+											case ServiceTypeKind.Decimal:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!, "M");
+												break;
+											}
```

Comment:
I understand there to be some implicit conversions in comparisons of numeric values (like int and long). Always adding the suffix seems more consistent to me, but let me know if you have a better way to do this.

Reply:
Seems reasonable.